htlcswitch/iterator.go

```diff
@@ -139,24 +145,81 @@ func (r *sphinxHopIterator) EncodeNextHop(w io.Writer) error {
 // hop to authenticate the information given to it by the prior hop.
 //
 // NOTE: Part of the HopIterator interface.
-func (r *sphinxHopIterator) ForwardingInstructions() ForwardingInfo {
-	fwdInst := r.processedPacket.ForwardingInstructions
-
-	var nextHop lnwire.ShortChannelID
-	switch r.processedPacket.Action {
-	case sphinx.ExitNode:
-		nextHop = exitHop
-	case sphinx.MoreHops:
-		s := binary.BigEndian.Uint64(fwdInst.NextAddress[:])
-		nextHop = lnwire.NewShortChanIDFromInt(s)
+func (r *sphinxHopIterator) ForwardingInstructions() (ForwardingInfo, error) {
+	var (
+		nextHop lnwire.ShortChannelID
+		amt     uint64
+		cltv    uint32
+	)
+
+	switch r.processedPacket.Payload.Type {
+	// If this is the legacy payload, then we'll extract the information
+	// directly from the pre-populated ForwardingInstructions field.
+	case sphinx.PayloadLegacy:
+		fwdInst := r.processedPacket.ForwardingInstructions
+
+		switch r.processedPacket.Action {
+		case sphinx.ExitNode:
+			nextHop = exitHop
+		case sphinx.MoreHops:
+			s := binary.BigEndian.Uint64(fwdInst.NextAddress[:])
+			nextHop = lnwire.NewShortChanIDFromInt(s)
+		}
+
+		amt = fwdInst.ForwardAmount
+		cltv = fwdInst.OutgoingCltv
+
+	// Otherwise, if this is the TLV payload, then we'll make a new stream
+	// to decode only what we need to make routing decisions.
+	case sphinx.PayloadTLV:
+		var cid uint64
+
+		tlvStream, err := tlv.NewStream(
+			tlv.MakeDynamicRecord(
+				tlv.AmtOnionType, &amt, nil,
+				tlv.ETUint64, tlv.DTUint64,
+			),
+			tlv.MakeDynamicRecord(
+				tlv.LockTimeOnionType, &cltv, nil,
+				tlv.ETUint32, tlv.DTUint32,
+			),
+			tlv.MakePrimitiveRecord(tlv.NextHopOnionType, &cid),
+		)
+		if err != nil {
+			return ForwardingInfo{}, err
+		}
+
+		err = tlvStream.Decode(bytes.NewReader(
+			r.processedPacket.Payload.Payload,
+		))
+		if err != nil {
+			return ForwardingInfo{}, err
```

Comment:
If I run the test "sendpayment with custom tlv record" again, I get `2019-08-15 10:29:32.757 [ERR] HSWC: unable to decode forwarding instructions: unknown required type: 100`. Probably originating here.

Reply:
That's expected as you're using an even type, which must be known with the way decoding works atm. Odd types should be used for cases where it isn't forwarding critical (custom extensions by developers).

Reply:
Ok clear. Test passes with type 101. When we are actually going to extract those custom values  and store them in the invoice db in a follow up pr, we'll do the tlv extension required for that, right?

Reply:
My previous comment about restricting the scope of this pr was based on the incorrect assumption that the error above needed tlv changes to be fixed.